### PR TITLE
Refactor/framework agnostic refactor

### LINKED ISSUE #224: Refactor: lift algorithm code into framework, remove PolicyExtras
## Summary

Lift the four duplicated algorithm implementations (CMA-ES, REINFORCE, LSTM-ES, DQN) and `ReplayBuffer` out of `games/tmnf/policies.py`, `games/sc2/policies.py`, and `games/sc2/sc2_policies.py` into shared `framework/` modules. Move the per-policy concerns (hyperparameter validation, resume-from-checkpoint, trainer-state load, loop dispatch) onto the policy classes themselves via class-level attributes (`POLICY_TYPE`, `LOOP_TYPE`, `VALID_POLICY_PARAMS`, `make` classmethod) and a `POLICY_REGISTRY` populated by an `@register_policy` decorator. With those in place, delete `PolicyExtras`, `GameAdapter.build_extras`, and the `extras` parameter on `train_rl`.

## Context

Today the repo carries ~5,000 lines across `games/tmnf/policies.py` (1,574), `games/sc2/policies.py` (1,720), and `games/sc2/sc2_policies.py` (1,783) that re-implement the same four algorithms. The framework already has the right shape for simple policies (`BasePolicy`, `WeightedLinearPolicy`, `NeuralNetPolicy`, `QTablePolicy`, `GeneticPolicy` in `framework/policies.py`) but never grew framework-level homes for the heavier algorithms.

To bridge the gap, the repo grew a `PolicyExtras` dataclass (`framework/run_config.py:96`) carrying:
- `factories: {policy_type → zero-arg callable}`
- `loop_dispatch: {policy_type → "hill_climbing"|"q_learning"|"cmaes"|"genetic"}`

Each adapter implements `build_extras(...)` (e.g. `games/tmnf/adapter.py:115`, `games/sc2/adapter.py:218`). The closures inside do three jobs the policy class itself should own:
1. Read/parse hyperparams out of `policy_params`.
2. Resume from `weights_file` via `from_cfg(...)`.
3. Load trainer state from `trainer_state.npz` with warn-on-failure.

SC2 also keeps a `_SC2_VALID_POLICY_PARAMS` table and `_validate_sc2_policy_config` (`games/sc2/adapter.py:40,75`) — schema knowledge that belongs on the policy class.

## Outcome

- Game `policies.py` files shrink to ~250 lines each (just per-game head encoders + small registration shells).
- `framework/training.py:_make_policy` becomes a single registry lookup.
- `GameAdapter` loses `build_extras`; `train_rl` loses its `extras` parameter.
- Net: ~−2,500 lines.

## Goals

1. One canonical implementation of `ReplayBuffer`, CMA-ES, REINFORCE, LSTM-ES, DQN in `framework/`.
2. Per-game policy files only contain code that is genuinely game-specific (action decoding, head structure, optional available-actions hook).
3. Remove `PolicyExtras` and the `extras` parameter on `train_rl`.
4. Remove `build_extras` from the `GameAdapter` protocol.
5. Preserve YAML weight-file compatibility (existing champions keep loading).
6. Preserve all current `policy_type` strings users put in `training_params.yaml` (with one documented exception: SC2 legacy bare-name policy_types — see Phase E).

## Non-goals

- Re-tuning hyperparameter defaults.
- Changing the public action/obs shapes of any game.
- Touching the simple framework policies beyond promoting them to the new registry pattern.

### New framework modules

| Module | Provides | Game-injected hooks |
|---|---|---|
| `framework/replay.py` | `ReplayBuffer`, `MaskedReplayBuffer` | none |
| `framework/dqn.py` | `DQNPolicy(BasePolicy)` — Adam MLP + target net + ε-greedy + replay | `discrete_actions: np.ndarray`, optional `available_actions_fn(info) → np.ndarray[bool]` |
| `framework/reinforce.py` | `REINFORCEPolicy(BasePolicy)` single softmax; `TwoHeadREINFORCEPolicy` for SC2's fn+spatial | `action_decoder(idx) → np.ndarray`, optional `available_actions_fn` |
| `framework/cmaes.py` | `CMAESDistribution` (pure Hansen-2016 math) + `CMAESPolicy(BasePolicy)` | `parameter_decoder(theta) → BasePolicy`, `flat_dim: int` |
| `framework/lstm.py` | `LSTMCore` + `LSTMEvolutionPolicy(BasePolicy)` isotropic-σ ES wrapper | `head_decoder(hidden) → np.ndarray`, optional `available_actions_fn` |

### Registry pattern (replaces PolicyExtras)

Added to `framework/policies.py:BasePolicy`:

```python
class BasePolicy(ABC):
    POLICY_TYPE: ClassVar[str] = ""
    LOOP_TYPE: ClassVar[str] = "hill_climbing"  # hill_climbing|q_learning|cmaes|genetic
    VALID_POLICY_PARAMS: ClassVar[frozenset[str]] = frozenset()

    @classmethod
    def make(cls, *, obs_spec, head_names, discrete_actions,
             weights_file, policy_params, re_initialize) -> "BasePolicy":
        cls._validate_params(policy_params)
        return cls._construct_or_resume(...)

POLICY_REGISTRY: dict[str, type[BasePolicy]] = {}

def register_policy(cls):
    if not cls.POLICY_TYPE:
        raise ValueError(f"{cls.__name__} must set POLICY_TYPE")
    if cls.POLICY_TYPE in POLICY_REGISTRY:
        raise ValueError(f"Duplicate policy_type {cls.POLICY_TYPE!r}")
    POLICY_REGISTRY[cls.POLICY_TYPE] = cls
    return cls
```

`_construct_or_resume` is the standard template that handles `re_initialize`, `from_cfg` resume, and `load_trainer_state` with warn-on-failure — currently repeated 11× across `games/sc2/adapter.py`.

### Training loop changes

`framework/training.py:_make_policy` collapses to a registry lookup; greedy-loop dispatch reads `policy.LOOP_TYPE` instead of consulting `extra_loop_dispatch`; `train_rl(...)` drops its `extras` parameter.

### Adapter + protocol changes

- `framework/game_adapter.py`: delete `build_extras` from the `GameAdapter` protocol.
- `framework/run_config.py`: delete the `PolicyExtras` dataclass.
- `main.py:222`: drop the `extras=adapter.build_extras(...)` argument.
- Each `games/*/adapter.py`: delete `build_extras`.

## Sub-issues (phased rollout — each PR keeps tests green)

Five PRs, in strict order. Each phase is blocked by the previous one. Sub-issues will be linked below as they are created.

1. **Phase A** — Add framework modules (no game changes).
2. **Phase B** — Add `POLICY_REGISTRY` + class-level `POLICY_TYPE`/`LOOP_TYPE`/`make` on `BasePolicy`; decorate framework built-ins. *(Blocked by A)*
3. **Phase C** — Migrate game policies to thin registered subclasses pinning decoders. *(Blocked by B)*
4. **Phase D** — Remove `PolicyExtras`, `build_extras`, and `extras`/`extra_policy_types`/`extra_loop_dispatch` plumbing. *(Blocked by C)*
5. **Phase E** — Delete SC2 legacy bare-name policy_types (silent breaking change); update CLAUDE.md + `tests/README.md`. *(Blocked by D)*

## Resolved decisions

1. **SC2 legacy bare-name policy_types** (`cmaes`, `reinforce`, `lstm`, `neural_dqn` on SC2 — already labelled *(legacy)* in CLAUDE.md) **are deleted silently** in Phase E. No deprecation alias, no runtime migration hint — the generic "Unknown policy_type" error from `_make_policy` (which lists registered types) is the only signal.
2. **Five staged PRs** (A → B → C → D → E), each independently reviewable with green tests at every boundary.

## Verification (whole refactor, asserted at Phase E)

1. `python -m pytest tests/` green.
2. `tests/test_policy_registry.py` (new) — every `policy_type` referenced in `config/**/*.yaml` and `games/*/config/*.yaml` resolves through `POLICY_REGISTRY` with a valid `LOOP_TYPE`.
3. Champion-file round-trip: load each `policy_type`'s YAML champion saved by pre-refactor code, train one sim, save, reload — `to_cfg()` byte-stable for linear cases, structurally equal for the others.
4. Smoke run: `python main.py smoke_refactor --game tmnf` with `policy_type: cmaes`, `n_sims: 2` — identical first-generation reward distribution as pre-refactor baseline (seeded).
5. SC2 smoke: `python main.py smoke_sc2 --game sc2` with `policy_type: sc2_genetic` — first-generation reward distribution matches baseline.
6. `grep -rn "PolicyExtras\|build_extras\|extra_policy_types\|extra_loop_dispatch"` returns no hits.

## Risks & mitigations

| Risk | Mitigation |
|---|---|
| Champion YAML files break (key layout drift in the decoder split) | Round-trip fixtures in Phase A; the decoder is the *inner* policy's existing `from_cfg`, so keys don't change. |
| SC2 user with bare `cmaes`/`reinforce`/`lstm`/`neural_dqn` in their config breaks | Documented breaking change in Phase E. CLAUDE.md already labels these "(legacy)". |
| Trainer-state `.npz` schema drift breaks resume | Framework-side `save_trainer_state` writes the same arrays as the per-game versions used (verified by porting the existing serialisation method verbatim). |
| Distributed worker calls `train_rl` differently | Search confirms `train_rl` is invoked from `main.py:222` only with `extras=`; verify in Phase D and update worker if needed. |
| Subtle math drift between three CMA-ES copies | Side-by-side numeric test on a fixed seed before deleting the per-game copy; identical Hansen-2016 constants confirmed at `games/tmnf/policies.py:712–722`, `games/sc2/policies.py:947–960`, `games/sc2/sc2_policies.py:1054–1062`. |
| `policies.py` repo-root shim breaks imports | Re-point the shim re-exports in the same PR; add a deprecation comment. |

## Estimated impact

- **Lines deleted:** ~3,500 (from `games/tmnf/policies.py`, `games/sc2/policies.py`, `games/sc2/sc2_policies.py`, `games/*/adapter.py`).
- **Lines added:** ~1,000 (framework modules ~800, registered subclasses across games ~200).
- **Net:** ~−2,500 lines.
- **Surface removed:** `PolicyExtras` dataclass, `GameAdapter.build_extras`, `train_rl(extras=...)`, `_make_policy(extra_policy_types=...)`, `extra_loop_dispatch` arg, `_SC2_VALID_POLICY_PARAMS`, `_validate_sc2_policy_config`.

### PR DESCRIPTION
**Pull request by espenhk**
Closes #224 

## Summary

Massive refactor to make policies more game-agnostic, making it easier to add new games. This one requires very thorough review.

## Related issues

#225 #228 #227 #231 #232 

## Validation

Changes have been independently reviewed, but need a big pass after resolving merge conflicts.

```bash
PYTHONPATH=. poetry run python -m pytest tests/ \
    --ignore=tests/test_env_termination.py \
    --ignore=tests/test_grid_search.py \
    --ignore=tests/integration/
```

## Bug fix details (delete this section if not a bug fix)

- Problem:
- Root cause:
- Fix:

## Feature details (delete this section if not a feature)

- User problem:
- Proposed solution:
- Trade-offs / follow-ups:

## Refactor / docs / chore details (delete this section if not applicable)

- What changed:
- Why this is safe:

### Release bump type

- [ ] Patch (default)
- [x] Minor
- [ ] Major

### AI usage

- [ ] No AI was used to write this code
- [X] AI was used to write/generate some or all of this code

**If AI was used:** Please describe which AI tool(s) and model(s) were used (e.g., Claude Opus, ChatGPT, etc.):

**Human involvement:**

Minimal. Each PR was sent through Copilot once, with fixing and then re-checking.

## Notes for the reviewer

- [ ] Updated `README.md` if user-visible behaviour, install steps, or CLI flags changed
- [ ] Updated `CLAUDE.md` if architecture, config knobs, or training-loop semantics changed
- [ ] Updated the relevant `games/<name>/README.md` for per-game changes
- [ ] Updated `tests/README.md` if tests were added, removed, or substantially changed (required by the repo's test-suite contract)
- [ ] Added an entry under `## [Unreleased]` in `CHANGELOG.md` for any user- or developer-visible change (new feature, new config key, breaking change, bug fix, dependency change). Trivial commits — experiment dumps, formatting, no-op refactors — can be skipped.
- [ ] New config keys appear in the relevant `config/*.yaml` master file with a sensible default
- [ ] Scope is limited to this issue/PR
- [ ] Tests updated where needed
- [ ] Docs updated where needed
- [ ] No secrets, large binaries, or experiment outputs committed

### CODE REVIEW

**Comment by espenhk** [comment]
@copilot resolve merge conflicts, then adress PR comments

**Comment by espenhk** [comment]
@copilot resolve merge conflicts

**Comment by espenhk** [comment]
@copilot+claude-opus-4.6 one final pass to make sure everything looks good after merge conflict resolution. Be super thorough.

**Comment by espenhk** [comment]
@copilot there are now many SC2-specific policies. How do these relate to framework? Are they just a workaround?
